fontique/src/backend/coretext.rs

```diff
@@ -75,97 +71,45 @@ impl SystemFonts {
     pub(crate) fn fallback(&mut self, key: impl Into<FallbackKey>) -> Option<FamilyId> {
         let key = key.into();
         let sample = key.script().sample()?;
-        self.fallback_for_text(sample, key.locale(), false)
+        let font = fallback_font_for_text(sample, key.locale(), false)?;
+        let family_name = unsafe { CTFontCopyFamilyName(&font) };
+        self.name_map.get(&family_name.to_string()).map(|n| n.id())
     }
 }
 
-impl SystemFonts {
-    fn fallback_for_text(
-        &mut self,
-        text: &str,
-        locale: Option<&str>,
-        prefer_ui: bool,
-    ) -> Option<FamilyId> {
-        let font = fallback_for_text(text, locale, prefer_ui)?;
-        self.name_map.get(&font.family_name()).map(|n| n.id())
+fn get_base_font(prefer_ui_font: bool) -> CFRetained<CTFont> {
+    if prefer_ui_font {
+        if let Some(font) =
+            unsafe { CTFontCreateUIFontForLanguage(CTFontUIFontType::System, 0.0, None) }
+        {
+            return font;
+        }
+    }
+    unsafe {
+        let attrs = CFDictionaryCreate(None, null_mut(), null_mut(), 0, null(), null());
+        let desc = CTFontDescriptorCreateWithAttributes(&attrs.unwrap());
+        CTFontCreateWithFontDescriptor(&desc, 0.0, null())
     }
 }
 
-fn fallback_for_text(text: &str, locale: Option<&str>, prefer_ui: bool) -> Option<CTFont> {
-    let cf_text = CFString::new(text);
-    let cf_text_range = CFRange::init(0, cf_text.char_len());
-    let cf_locale = locale.map(CFString::new);
-    let base_font = {
-        let desc_attrs = CFDictionary::from_CFType_pairs(&[]);
-        let mut desc = font_descriptor::new_from_attributes(&desc_attrs);
-        if prefer_ui {
-            if let Some(ui_desc) = ui_font_for_language(kCTFontSystemFontType, 0.0, None)
-                .map(|font| font.copy_descriptor())
-            {
-                desc = ui_desc;
-            }
-        }
-        font::new_from_descriptor(&desc, 0.0)
+fn fallback_font_for_text(
+    text: &str,
+    locale: Option<&str>,
+    prefer_ui_font: bool,
+) -> Option<CFRetained<CTFont>> {
+    let text = CFString::from_str(text);
+    let text_range = CFRange {
+        location: 0,
+        length: unsafe { CFStringGetLength(&text) },
```

Comment:
Out of interest, is this the same as (before shadowing) `text.chars().count()`?

Reply:
I honesty don't know.

Reply:
I *think* it would be: `text.chars().map(char::len_utf16).sum()` ([docs](https://doc.rust-lang.org/stable/std/primitive.char.html#method.len_utf16))